system/lib/wasmfs/syscalls.cpp

```diff
@@ -719,4 +719,142 @@ long __syscall_getdents64(long fd, long dirp, long count) {
 
   return bytesRead;
 }
+
+long __syscall_rename(long old_path, long new_path) {
+  // This syscall must acquire locks on the old_path parent and new_path parent for atomicity.
+  // This is to prevent other file system operations from changing the state during rename.
+  // For example, two renames could be contending for the same source and destination.
+  // 1. Obtain the old path parent first and lock it.
+  // 2. Obtain the old path child.
+  // 3. Try lock the new path parent. Operation should return EBUSY if it is
+  // already locked. In the case where the old path parent and new path parent are the
+  // same, a recursive lock should allow it to be locked again from the same
+  // thread.
+  // Renaming a file:
+  // 4. a) If the old path child is a file, ensure the new path child is not a
+  // directory.
+  // 4. b) If the new path child does not exist, add the old path child to the
+  // new path parent.
+  // 4. c) If the new path child does exist, unlink the new path child and add
+  // the old path child. 
+  // Renaming a directory:
+  // 5. a) If the old path child is a directory, ensure the new path child is
+  // not a file.
+  // 5. b) If the new path child does not exist, do the same as 4b.
+  // 5. c) If the new path child does exist, ensure the newpath child is empty
+  // and then do 4c.
+
+  // If old_path and new_path are the same, do nothing.
+  if (strcmp((char*)old_path, (char*)new_path) == 0) {
+    return 0;
+  }
+
+  auto oldPathParts = splitPath((char*)old_path);
+
+  if (oldPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming the root directory returns EBUSY.
+  // TODO: Fix this when path parsing is refactored.
+  if (oldPathParts.size() == 1 && oldPathParts[0] == "/") {
+    return -EBUSY;
+  }
```

Comment:
This needs to wait for the path parsing refactoring. If we changed it now it would conflict with the case where `parent` == `root` and `child` == `null`.